keystore/v2/keystore/filesystem/keyRing.go

```diff
@@ -0,0 +1,255 @@
+func (r *KeyRing) AddKey(key api.KeyDescription) (api.MutableKey, error) {
+	k, data, err := newKey(r, key)
+	if err != nil {
+		r.log.WithError(err).Debug("failed to make new key")
+		return nil, err
+	}
+	err = r.addKey(k, data)
+	if err != nil {
+		r.log.WithError(err).Debug("failed to add new key")
+		return nil, err
+	}
+	r.log.Trace("new key added to key ring")
+	return k, nil
+}
+
+//
+// Key data encryption
+//
+
+func (r *KeyRing) keyRingContext(context []byte) []byte {
+	c := make([]byte, 0, len("key ring ")+len(r.path)+len(": ")+len(context))
+	c = append(c, "key ring "...)
+	c = append(c, r.path...)
+	c = append(c, ": "...)
+	c = append(c, context...)
+	return c
+}
+
+func (r *KeyRing) encrypt(data, context []byte) ([]byte, error) {
+	return r.store.encrypt(data, r.keyRingContext(context))
+}
+
+func (r *KeyRing) decrypt(data, context []byte) ([]byte, error) {
+	return r.store.decrypt(data, r.keyRingContext(context))
+}
+
+//
+// Transaction handling
+//
+
+func (r *KeyRing) pendingUpdates() bool {
+	return len(r.txLog) > 0
+}
+
+func (r *KeyRing) applyPendingTX() error {
+	for lastTX, tx := range r.txLog {
+		err := tx.Apply(r)
+		if err != nil {
+			r.log.WithError(err).Debug("failed to apply update")
+			// Apply damage control. We cannot handle double faults so our best option
+			// is to complain in logs and then keep calm and carry on, betting on luck.
+			// Rollback should not fail, actually, but panicking is too destructive now.
+			// TODO: is it possible to just snapshot the original state instead?
+			for i := lastTX - 1; i >= 0; i-- {
+				err := r.txLog[i].Rollback(r)
+				if err != nil {
+					r.log.WithError(err).Warn("failed to roll back update")
+				}
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *KeyRing) commitTX() {
+	r.txLog = nil
+}
+
+func (r *KeyRing) setCurrent(newSeqnum int) error {
+	oldSeqnum := asn1.NoKey
+	if r.current != nil {
+		oldSeqnum = r.current.seqnum
+	}
+	r.pushTX(&txSetKeyCurrent{oldSeqnum, newSeqnum})
+	err := r.store.syncKeyRing(r)
+	if err != nil {
+		r.popTX()
+	}
+	return err
+}
+
+func (r *KeyRing) changeKeyState(keySeqnum int, oldState, newState api.KeyState) error {
+	r.pushTX(&txChangeKeyState{keySeqnum, oldState, newState})
+	err := r.store.syncKeyRing(r)
+	if err != nil {
+		r.popTX()
+	}
+	return err
+}
+
+func (r *KeyRing) addKey(newKey *Key, newData *asn1.Key) error {
+	if newKey.ring != r {
+		panic("mismatched key ring")
+	}
+	r.pushTX(&txAddKey{newKey, newData})
+	err := r.store.syncKeyRing(r)
+	if err != nil {
+		r.popTX()
+	}
+	return err
+}
+
+//
+// Internal utilities
+//
+
+func (r *KeyRing) keyDataBySeqnum(seqnum int) *asn1.Key {
+	key, _ := r.data.KeyWithSeqnum(seqnum)
+	return key
+}
+
+// Don't start with zero to not match zero-initialized defaults.
+const firstSeqnum = 1
+
+func (r *KeyRing) nextSeqnum() int {
+	if len(r.data.Keys) == 0 {
+		return firstSeqnum
+	}
+	return r.data.Keys[len(r.data.Keys)-1].Seqnum + 1
```

Comment:
the newest key at the end? what if we will sort by some other logic our keys (new first, destroyed last) and their numbers will not sorted by ascending? what about to cache max(r.data.Keys.Seqnum...) value on load keyring from storage and use it? or if we will use not array struct but some linked list or tree. then retrieve last key wiill be not such easy

Reply:
> what if we will sort by some other logic our keys (new first, destroyed last) and their numbers will not sorted by ascending? what about to cache max(r.data.Keys.Seqnum...) value on load keyring from storage and use it?

Good point! Yes, current implementation assumes that the keys are stored with ascending sequence numbers. This assumption breaks if we start using a different order.

It’s a good idea to just store the last seqnum in key ring directly. They should not be reused. Currently we do not implement key removal, but imagine that we have this:

```
[1, 2, 3, 4, 5]
```

then the key number 5 gets removed. If we don’t record that the next seqnum should be 6, we will reuse 5 and much confusion will arise (say, because someone tracks hashes of keys tied to their seqnums).

Reply:
I think that it good requirement and assumption that we will not physically remove any key. We can change state to "destroyed", store empty key material (empty or zeroized array) but leave metadata as is (time period, type, ...) for history and any audit requirements

Reply:
Yeah, that was the idea behind not allowing to remove keys from a key ring: mark them as destroyed instead. Though, I could imagine some pathological use cases where a key is frequently rotated and we end up with 1,000,000 useless destroyed keys. It would be nice to allow to prune metadata for older keys. That is possible to do without disturbing the currently active key (and hence the next seqnum). So as long as we don’t allow to remove the latest key, we don’t need to store the last used seqnum explicitly.

Reply:
> or if we will use not array struct but some linked list or tree. then retrieve last key wiill be not such easy

ASN.1 does not support linked lists directly and here `r.data.Keys` refers to parsed ASN.1. So if we assume that the keys are sorted by seqnum and active keys go last then we can get away with this implementation.

It allows some leeway, such as to put “unusable” keys first, followed by active ones, grouped by their state. For example, we can store keys like this:

```
Index | Seqnum | State      
------+--------+------------
    0 |      4 | compromised
    1 |      7 | compromised
    2 |      1 | deactivated
    3 |      2 | deactivated
    4 |      3 | deactivated
    5 |      5 | active     
    6 |      6 | active     
    7 |      8 | pre-active 
    8 |      9 | pre-active 
```

This makes it cheap to add new keys and relatively cheap to filter for “active” keys. You can get cheap addition by storing last seqnum separately, but if the keys are sorted by seqnum alone you still need to look through all of them to filter by state.